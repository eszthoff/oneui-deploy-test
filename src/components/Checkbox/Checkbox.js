import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './Checkbox.scss';
import { CHECKBOX_VIEWBOX } from '../../constants';

const { block, elem } = bem({
    name: 'Checkbox',
    classnames: styles,
    propsToMods: ['disabled']
});

const Checkbox = props => {
    const { id, children, viewbox, disabled, ...rest } = props;

    return (
        <label {...block(props)} htmlFor={id} disabled={disabled}>
            <input
                type="checkbox"
                id={id}
                {...rest}
                {...elem('input', props)}
                disabled={disabled}
            />
            <span {...elem('box', props)}>
                <svg {...elem('svg', props)} width="12px" height="10px" viewBox={viewbox}>
                    <polyline points="1.5 6 3.5 9 8 3" />
                </svg>
            </span>
            <span {...elem('text', props)}>{children}</span>
        </label>
    );
};

Checkbox.propTypes = {
    /** A unique id to reference this checkbox */
    id: PropTypes.string.isRequired,
    /** If the checkbox should be disabled */
    disabled: PropTypes.bool,
    /** The label for the checkbox */
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    /** Viewbox dimentions, allows to costumise the checkbox */
    viewbox: PropTypes.arrayOf(PropTypes.number)
};

Checkbox.defaultProps = {
    disabled: false,
    viewbox: CHECKBOX_VIEWBOX
};

export default Checkbox;
