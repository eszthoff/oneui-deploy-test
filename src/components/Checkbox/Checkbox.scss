@import '../../themes/nice/nice.scss';

.Checkbox {
    margin: auto;
    user-select: none;
    cursor: pointer;
    display: block;

    &__text {
        display: inline-block;
        vertical-align: middle;
        transform: translate3d(0,0,0);
        padding: var(--spacing-normal) 0 var(--spacing-normal) var(--spacing-2x);
        color: var(--color-foreground);
    }

    &__box {
        display: inline-block;
        vertical-align: middle;
        box-sizing: content-box;
        position: relative;
        width: var(--font-size-small);
        height: var(--font-size-small);
        border-radius: var(--border-radius);
        transform: scale(1);
        border: 1px solid var(--color-neutral);
        transition: all var(--transition-duration) ease;

        &:before {
            content: "";
            width: 100%;
            height: 100%;
            background: var(--color-brand);
            display: block;
            transform: scale(0);
            opacity: 1;
            border-radius: 50%;
        }
    }

    &__svg {
        position: absolute;
        top: 0;
        left: 0;
        fill: none;
        stroke: var(--color-background);
        stroke-width: 2;
        stroke-linecap: round;
        stroke-linejoin: round;
        stroke-dasharray: var(--font-size-large);
        stroke-dashoffset: var(--font-size-large);
        transition: all var(--transition-duration) ease;
        transition-delay: .1s;
        transform: translate3d(0, 0, 0);
    }

    &__input {
        display: none;

        &:checked {
            + .Checkbox__box {
                background: var(--color-brand);
                border-color: var(--color-brand);
                animation: wave var(--transition-duration) ease;

                .Checkbox__svg {
                    stroke-dashoffset: 0;
                }

                &:before {
                    transform: scale(2.5);
                    opacity: 0;
                    transition: all var(--transition-duration) ease;
                }
            }
        }
    }

    &:hover {
        .Checkbox__box {
            border-color: var(--color-brand-60);
        }
    }

    &--disabled {
        cursor: not-allowed;
        .Checkbox__text {
            color: var(--color-muted);
        }

        &:hover {
            .Checkbox__box {
                border-color: var(--color-neutral);
            }
    
        }
    }
}
