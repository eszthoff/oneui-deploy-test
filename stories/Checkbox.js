import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import { boolean, text, withKnobs } from '@storybook/addon-knobs'; // eslint-disable-line import/no-extraneous-dependencies
import Checkbox from '../src/components/Checkbox';

storiesOf('Checkbox', module)
    .addDecorator(withKnobs)
    .add('Checkbox', () => (
        <Checkbox
            disabled={boolean('Disabled', false)}
            id={text('Id', 'checkbox-1')}
        >
            {text('Checkbox label', 'Select me!')}
        </Checkbox>
    ))
    .add('Checkbox with callback', () => (
        <Checkbox
            disabled={boolean('Disabled', false)}
            id={text('Id', 'checkbox-1')}
            onChange={() => alert('clicked')} // eslint-disable-line no-undef, no-alert
        >
            {text('Checkbox label', 'Select me!')}
        </Checkbox>
    ));
