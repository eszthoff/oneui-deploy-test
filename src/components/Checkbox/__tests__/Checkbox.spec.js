import React from 'react';
import toJson from 'enzyme-to-json';
import Checkbox from '../Checkbox';

describe('<Checkbox> that renders a checkbox', () => {
    it('should render default checkbox correctly', () => {
        const wrapper = shallow(<Checkbox id="c1">Check this out</Checkbox>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should add classes when props are changed', () => {
        const wrapper = shallow(
            <Checkbox id="c2" disabled>
                Useless checkbox
            </Checkbox>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
