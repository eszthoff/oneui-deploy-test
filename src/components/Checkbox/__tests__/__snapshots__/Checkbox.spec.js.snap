// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> that renders a checkbox should render default checkbox correctly 1`] = `
<Checkbox
  disabled={false}
  id="c1"
>
  <div
    className="Checkbox"
  >
    <input
      className="Checkbox__input"
      disabled={false}
      id="c1"
      type="checkbox"
    />
    <label
      className="Checkbox__label"
      disabled={false}
      htmlFor="c1"
    >
      <span
        className="Checkbox__box"
      >
        <svg
          className="Checkbox__svg"
          height="10px"
          viewBox={
            Array [
              0,
              0,
              10,
              10,
            ]
          }
          width="12px"
        >
          <polyline
            points="1.5 6 3.5 9 8 3"
          />
        </svg>
      </span>
      <Text
        className="Checkbox__text"
        inline={true}
        muted={false}
      >
        <span
          className="Text Checkbox__text"
          muted={false}
        >
          Check this out
        </span>
      </Text>
    </label>
  </div>
</Checkbox>
`;
